test/hamiltonian.jl

```diff
@@ -68,3 +69,32 @@ end
         end
     end
 end
+
+@testset "Energy with ComponentArrays" begin
+    n_tests = 10
+    for T in [Float32, Float64]
+        for _ in 1:n_tests
+            θ_init = ComponentArray(; a=randn(T, D), b=randn(T, D))
+            r_init = ComponentArray(; a=randn(T, D), b=randn(T, D))
+
+            h = Hamiltonian(UnitEuclideanMetric(T, 2 * D), ℓπ, ∂ℓπ∂θ)
+            @test -AdvancedHMC.neg_energy(h, r_init, θ_init) == sum(abs2, r_init) / 2
+            @test AdvancedHMC.∂H∂r(h, r_init) == r_init
+            @test typeof(AdvancedHMC.∂H∂r(h, r_init)) == typeof(r_init)
+
+            M⁻¹ = ones(T, 2 * D) + abs.(randn(T, 2 * D))
+            h = Hamiltonian(DiagEuclideanMetric(M⁻¹), ℓπ, ∂ℓπ∂θ)
+            @test -AdvancedHMC.neg_energy(h, r_init, θ_init) ≈
+                r_init' * diagm(0 => M⁻¹) * r_init / 2
+            @test AdvancedHMC.∂H∂r(h, r_init) == M⁻¹ .* r_init
+            @test typeof(AdvancedHMC.∂H∂r(h, r_init)) == typeof(r_init)
+
+            m = randn(T, 2 * D, 2 * D)
+            M⁻¹ = m' * m
```

Comment:
Isn't this the (only?) problem - if you're working with ComponentArrays, the Hamiltonian should also be constructed with a ComponentArray as matrix (with the correct axes), to exactly avoid this matrix multiplication issue? IMO users are just lucky that the other examples work (due to special broadcasting behaviour) but in principle also the DiagEuclideanMetric should be constructed with a ComponentArray (since equivalently - and probably without performance regression - one could apply it by multiplication with a `Diagonal` matrix).

So I actually think it seems to be a user issue and the only problem is that the Hamiltonian is not defined correctly.

Reply:
> thinking of mass matrix being a ComponentMatrix is bit of weird

I'd disagree, I'd say that's how you have to think about your mass matrix if your samples etc are ComponentArrays. For instance, in the example here there's an `a` and a `b` component of theta and r. So also the mass matrix (be it dense, diagonal or uniform) needs to consist of `a` and `b` components.

So I'd argue, instead of "fixing" the dense case an error should be thrown for the diagonal (and unit?) case as well. For instance, maybe it would be possible to check upfront that the axes of the mass matrices match the axis of sample, momentum etc and throw a descriptive error otherwise that makes it easier for users to realize that their Hamiltonian is incorrect.

Reply:
Fair point, I will add these axes checks and warnings when constructing Hamiltonian with ComponnetArrays.jl

Reply:
Okay, AdvancedHMC now has axes checks and warnings when dealing with ComponentArrays